File: tests/fixtures/typescript/src/store.ts
/** Types the service annotates against. */

export type Identifier = number;

export interface Labeled {
  describe(): string;
}

export class Base implements Labeled {
  label: string = "base";

  describe(): string {
    return this.label;
  }
}

export class Engine extends Base {
  region: string;

  constructor(region: string) {
    super();
    this.region = region;
  }

  ping(): boolean {
    return this.region.length > 0;
  }
}

File: tests/fixtures/rust/src/store.rs
//! Types the service annotates against.

/// An alias the service uses in signatures.
pub type Identifier = u32;

/// Implemented by Engine.
pub trait Labeled {
    fn describe(&self) -> String;
}

/// Carries a field.
pub struct Base {
    pub label: String,
}

impl Labeled for Base {
    fn describe(&self) -> String {
        self.label.clone()
    }
}

/// Embeds a base and adds a field.
pub struct Engine {
    pub base: Base,
    pub region: String,
}

impl Engine {
    pub fn new(region: String) -> Self {
        Self { base: Base { label: String::new() }, region }
    }

    pub fn ping(&self) -> bool {
        !self.region.is_empty()
    }
}
